### PR TITLE
v2pt_theory: Added zero velocity constraint requirement.

### PR DESCRIPTION
**Pull request by angadhn**
I'm unsure if what I am adding here, which is a trivial detail, was
intentionally left out; here is the description of what I have done:
P and O are two points of a rigid body B that is moving in frame N.
The velocity of P in N, N^v^P, can be computed using v2pt_theory,
which it does admirably. However, the function should also set
B^v^P = 0; that is what I have added. In fact, the function should
likely also constrain the B^v^O to be 0; that is what makes P and O
2 points of the same body.
Further, this constraint of B^v^O should also likely be added to v1pt
theory. If you agree, I can commit changes to that next time, as well.

<!-- If this pull request fixes an issue please indicate which issue by typing "Fixes #NNNN" below.>

### CODE REVIEW

**Comment by angadhn** [comment]
@moorepants Just wondering if this was not done for a reason. this == what is explained in my commit.

**Comment by angadhn** [comment]
@sympy/mechanics Thoughts?

**Comment by oliverlee** [comment]
:+1: I looked through my Kane examples and it looks like I had to explicitly set the velocity of the point in the fixed frame to zero.

However, please fix misspellings in the commit message and in the title of this pull request.

**Comment by moorepants** [comment]
Please add a unit test too if possible.

**Comment by angadhn** [comment]
I appended a couple lines to the tests for v1pt_theory and v2pt_theory which show that the zero velocity constrain is working. I believe the failed checks are unrelated to the recent commit but an extra set of eyes to verify that would be great.

**Comment by czgdp1807** [comment]
@sympy/mechanics Please provide comments on this PR so that this can be concluded.

**Comment by moorepants** [comment]
I think this is fine, but it would be helpful if the docstring for the two functions tell the user that this is what happens. It should explicitly say that 3 velocities are set to a value.

**Comment by czgdp1807** [comment]
Hi, @angadhn 
Thanks for the contribution. Are you still interested to continue work on this PR? Please address https://github.com/sympy/sympy/pull/12322#issuecomment-536695517

**Comment by czgdp1807** [comment]
@moorepants I am a bit unfamiliar with this module but I think this PR can be extended. It would be helpful if you can suggest the exact change(diff or docstring that you want) so that I can make a new PR to complete this. Thanks. 
PS - If someone is interested in this then please comment below and complete this.

**Comment by sachin-4099** [comment]
Hi, I am interested, please guide me as to what to do, 
I am not getting any clue as to what exactly needs to be done from the above conversations.

File: sympy/physics/vector/point.py [hunk 1]
```diff
@@ -380,6 +380,7 @@ def v1pt_theory(self, otherpoint, outframe, interframe):
         v1 = self.vel(interframe)
         v2 = otherpoint.vel(outframe)
         omega = interframe.ang_vel_in(outframe)
+        otherpoint.set_vel(interframe, 0)
```
Comment (moorepants) [inline sympy/physics/vector/point.py]: Currently, the user has to ensure that `otherpoint` is fixed in `interframe` before calling `a2pt_theory()`. But users can pick the wrong point by accident, thinking it was fixed. I think it would be better to warn the user that `otherpoint` is not fixed in `interframe` at this location in the code. Overwriting this velocity will lead to confusion by users.

**Comment by czgdp1807** [comment]
@sachin-4099 Would you like to take this over?

**Comment by sachin-4099** [comment]
Sorry, I am no longer interested.

File: sympy/physics/vector/point.py (hunk 1 above)
Comment (angadhn) [reply sympy/physics/vector/point.py:383]: Hi @moorepants, teaching this to my first year undergraduate students and this is definitely leading to a world of confusion for them that the `otherpoint` is not of zero velocity in the body that it is on. The 2 point theory is for two points on the same rigid body and therefore the `otherpoint` must have zero velocity in the `interframe`. So, I do not think that overwriting this velocity leads to confusion- if anything, I would argue that it leads to clarity that could be used for better bookkeeping of different points on a body.

Comment (moorepants) [reply sympy/physics/vector/point.py:383]: I just looked at this again to try to remember what it is about. I still have a similar concern. Right now, with most everything in the library, the user has to explicitly set the velocity of every point. The existing purpose of these velocity methods are really just helpers in calculating the velocity of a single point, it's never been expected that they establish velocities of multiple points. So, I'm still -1 to making this change without some examples of why it is necessary. There are lots of things that aren't set automatically for the user. We can of course start changing that, but we need to have good reason, provide examples of the necessity, and then document carefully any new behavior. Maybe start by providing an example problem where the current form of these methods causes issues. I personally haven't come across that yet, but maybe you are solving different problems or in a different way.

Comment (angadhn) [reply sympy/physics/vector/point.py:383]: Example is here: https://nbviewer.jupyter.org/github/angadhn/v1pt_changes/blob/main/Particle_on_Double_Pendulum.ipynb

TL; DR: velocities are not accessible for a points in all different frames.

Comment (moorepants) [reply sympy/physics/vector/point.py:383]: Thanks for the example. I see what you are expecting now. But you don't show why you need the velocities of those points set for any future steps in the problem. Knowing what necessitates these being set automatically could help me get on board.

The physics.vector design philosophy is effectively "every angular and linear velocity must be explicitly set by the user" and, as it currently stands, every command in sympy.physics.vector sets only a single velocity. There isn't the concept (yet) of an all-knowing system that is auto-magically making sure you do all your kinematics correctly. In fact, I think that is a feature, especially for learning. You want your students to control the specification of the positions, velocities, and accelerations and their relationships explicitly when learning. If we do more than that, we move towards a physics simulator that abstracts those details that are needed when teaching dynamics. For example, in the grad dynamics course I show them how to manually calculate v1/2pt relationships, then later tell them that there is this method that saves you a bit of typing.

For your example, if you need all those velocities set, this is the expected thing to type:

```python
O.set_vel(A, 0)
Q.set_vel(A, 0)
Q.v2pt_theory(O, N, A)
```

Every velocity is explicitly set by the user.  Each of those commands sets a single velocity.

I understand the desire to not have to type everything explicitly, but we've been trying to approach that by building on top of what's there (for example see the open pr's about adding a joint-type, which sets numerous velocities in one call).

The change, as proposed here, changes 10+ years of expected behavior from these methods and sets velocities that have never been set before. So I'm still not keen on merging this as proposed. One solution is to teach students how to make their own functions:

```python
def set_all_vel_with_v2pt(O, Q,  N, A):
    O.set_vel(A, 0)
    Q.set_vel(A, 0) 
    Q.set_vel(N, O.vel(N) + cross(A.ang_vel(N), Q.pos_from(O))
```
Then they write the function and know exactly what it does when they use it. This treats physics.vector as the base building block for the students to develop their proper kinematics methods.

Comment (moorepants) [reply sympy/physics/vector/point.py:383]: One more point is that these methods currently do exactly what they say in the docstring. The docstrings say anything about setting velocities on the other points passed in to the method. But maybe, there simply needs to be more clarity in the docstring of what the method does. Could that reduce confusion for your students?

Comment (moorepants) [reply sympy/physics/vector/point.py:383]: I chatted with @angadhn on the phone about this PR. I wanted to add somethings to look into:

- In the notebook he shared the last line is `O.vel(B)` which was never set explicitly yet it returns a value and we should check if that value is correct. It could be that this new change in #20049 causes that to be calculated.
- I don't think there is any precedence in the existing methods to set more than one velocity on any given method call on Point or ReferenceFrame. It has always been the case that calling `A.ang_vel_in(N)` will search the orientation connections between frames A and N and if there are rotation matrices fully connecting both, that the velocity will be calculated if it was never set. We very recently added that same feature, to be consistent, in PR #20049 for Points. But I don't think any intermediate velocities are set, only the single one that is requested. We should double check this though.

**Comment by moorepants** [comment]
In the notebook shared above this is the code:

```
In [1]: from sympy import symbols, sin, cos

In [2]: from sympy.physics.mechanics import dynamicsymbols, ReferenceFrame, init_vprinting

In [3]: l = symbols('l')

In [4]: theta_1, theta_2, x = dynamicsymbols('theta_1, theta_2, x')

In [5]: x_dot = dynamicsymbols('x', 1)

In [6]: N = ReferenceFrame('N')

In [7]: A = ReferenceFrame('A')

In [8]: B = ReferenceFrame('B')

In [9]: A.orient(N, 'Axis', (theta_1, N.z))

In [10]: A.ang_vel_in(N)
Out[10]: Derivative(theta_1(t), t)*N.z

In [11]: B.orient(A, 'Axis', (theta_2, A.z))

In [12]: B.ang_vel_in(A)
Out[12]: Derivative(theta_2(t), t)*A.z

In [13]: from sympy.physics.mechanics import Point

In [14]: O = Point('O')

In [15]: Q = Point('Q')

In [16]: P = Point('P')

In [17]: Q.set_pos(O, l*A.y)

In [18]: Q.pos_from(O)
Out[18]: l*A.y

In [19]: O.set_vel(N, 0)

In [20]: Q.v2pt_theory(O, N, A)
Out[20]: - l*Derivative(theta_1(t), t)*A.x

In [21]: P.set_pos(Q, x*B.y)

In [22]: P.pos_from(Q)
Out[22]: x(t)*B.y

In [23]: P.set_vel(B, x_dot*B.y)

In [24]: P.vel(B)
Out[24]: Derivative(x(t), t)*B.y

In [25]: P.v1pt_theory(Q,N,B)
Out[25]: - l*Derivative(theta_1(t), t)*A.x - (Derivative(theta_1(t), t) + Derivative(theta_2(t), t))*x(t)*B.x + Derivative(x(t), t)*B.y

In [26]: O.set_vel(A,0)

In [27]: O.vel(B)
Out[27]: - l*Derivative(theta_2(t), t)*A.x
```

One question we discussed was whether line 27 gives correct output. In this case, The velocity of point O was never explicitly set in reference frame B. Prior to SymPy 1.7, line 27 would have returned `ValueError: Velocity of point O has not been defined in ReferenceFrame B`. But with PR #20049, it now calculates that velocity if there is enough position and/or velocity information present. This was introduced to provide consistent behavior as when you do this with reference frames, i.e. ask for an angular velocity when it hasn't been explicitly set. I believe the output on [27] is correct in this case.

Also note that only O in B was set:

```
In [28]:  O._vel_dict
Out[28]: {N: 0, A: 0, B: - l*Derivative(theta_2(t), t)*A.x}
```

It only sets the one velocity. It doesn't set O in A or O in N unless you ask for those.

**Comment by moorepants** [comment]
Pasting this here for later reference to see when there is actually enough information in place to calculate `O.vel(B)`:

```
In [1]: import sympy

In [2]: sympy.__version__
Out[2]: '1.8.dev'

In [3]: %paste
In [1]: from sympy import symbols, sin, cos

In [2]: from sympy.physics.mechanics import dynamicsymbols, ReferenceFrame, init_vprinting

In [3]: l = symbols('l')

In [4]: theta_1, theta_2, x = dynamicsymbols('theta_1, theta_2, x')

In [5]: x_dot = dynamicsymbols('x', 1)

In [6]: N = ReferenceFrame('N')

In [7]: A = ReferenceFrame('A')

In [8]: B = ReferenceFrame('B')

In [9]: A.orient(N, 'Axis', (theta_1, N.z))

In [10]: A.ang_vel_in(N)

In [11]: B.orient(A, 'Axis', (theta_2, A.z))

In [12]: B.ang_vel_in(A)

In [13]: from sympy.physics.mechanics import Point

In [14]: O = Point('O')

In [15]: Q = Point('Q')

In [16]: P = Point('P')

In [17]: Q.set_pos(O, l*A.y)

In [18]: Q.pos_from(O)
## -- End pasted text --
Out[3]: l*A.y

In [4]: O.vel_dict
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-4-68b42d933d60> in <module>
----> 1 O.vel_dict

AttributeError: 'Point' object has no attribute 'vel_dict'

In [5]: O._vel_dict
Out[5]: {}

In [6]: O.set_vel(N, 0)

In [7]: O._vel_dict
Out[7]: {N: 0}

In [8]: Q.v2pt_theory(O, N, A)
Out[8]: - l*Derivative(theta_1(t), t)*A.x

In [9]: O._vel_dict
Out[9]: {N: 0}

In [10]: O.set_vel(A)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-10-fcf5ab42bda6> in <module>
----> 1 O.set_vel(A)

TypeError: set_vel() missing 1 required positional argument: 'value'

In [11]: O.set_vel(A, 0)

In [12]: O._vel_dict
Out[12]: {N: 0, A: 0}

In [13]: O.vel(B)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-13-31184ba45e73> in <module>
----> 1 O.vel(B)

~/src/sympy/sympy/physics/vector/point.py in vel(self, frame)
    567             else:
    568                 raise ValueError('Velocity of point ' + self.name + ' has not been'
--> 569                              ' defined in ReferenceFrame ' + frame.name)
    570 
    571         return self._vel_dict[frame]

ValueError: Velocity of point O has not been defined in ReferenceFrame B

In [14]: O._vel_dict
Out[14]: {N: 0, A: 0}

In [15]: P.v1pt_theory(Q,N,B)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-15-c9aadb955a36> in <module>
----> 1 P.v1pt_theory(Q,N,B)

~/src/sympy/sympy/physics/vector/point.py in v1pt_theory(self, otherpoint, outframe, interframe)
    443         _check_frame(interframe)
    444         self._check_point(otherpoint)
--> 445         dist = self.pos_from(otherpoint)
    446         v1 = self.vel(interframe)
    447         v2 = otherpoint.vel(outframe)

~/src/sympy/sympy/physics/vector/point.py in pos_from(self, otherpoint)
    303 
    304         outvec = Vector(0)
--> 305         plist = self._pdict_list(otherpoint, 0)
    306         for i in range(len(plist) - 1):
    307             outvec += plist[i]._pos_dict[plist[i + 1]]

~/src/sympy/sympy/physics/vector/point.py in _pdict_list(self, other, num)
    113             return outlist[0]
    114         raise ValueError('No Connecting Path found between ' + other.name +
--> 115                          ' and ' + self.name)
    116 
    117     def a1pt_theory(self, otherpoint, outframe, interframe):

ValueError: No Connecting Path found between Q and P

In [16]: P.set_pos(Q, x*B.y)

In [17]: P.v1pt_theory(Q,N,B)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-17-c9aadb955a36> in <module>
----> 1 P.v1pt_theory(Q,N,B)

~/src/sympy/sympy/physics/vector/point.py in v1pt_theory(self, otherpoint, outframe, interframe)
    444         self._check_point(otherpoint)
    445         dist = self.pos_from(otherpoint)
--> 446         v1 = self.vel(interframe)
    447         v2 = otherpoint.vel(outframe)
    448         omega = interframe.ang_vel_in(outframe)

~/src/sympy/sympy/physics/vector/point.py in vel(self, frame)
    567             else:
    568                 raise ValueError('Velocity of point ' + self.name + ' has not been'
--> 569                              ' defined in ReferenceFrame ' + frame.name)
    570 
    571         return self._vel_dict[frame]

ValueError: Velocity of point P has not been defined in ReferenceFrame B

In [18]: P.set_vel(B, x_dot*B.y)

In [19]: P.v1pt_theory(Q,N,B)
Out[19]: - l*Derivative(theta_1(t), t)*A.x - (Derivative(theta_1(t), t) + Derivative(theta_2(t), t))*x(t)*B.x + Derivative(x(t), t)*B.y

In [20]: O._vel_dict
Out[20]: {N: 0, A: 0}

In [21]: O.vel(B)
Out[21]: - l*Derivative(theta_2(t), t)*A.x

In [22]: O._vel_dict
Out[22]: {N: 0, A: 0, B: - l*Derivative(theta_2(t), t)*A.x}
```

Seems like P.v1pt_theory and it's associated setup supplies enough info for the automatic cacl of O in B. Will have to dig a bit more later.

**Comment by TJStienstra** [comment]
When going through this issue I would summarize the problem as the following. The velocity theorems both assume the reference point (`O`) to be fixed in the interframe (`B`). A propose solution was to set the velocity of `O` to `0` in `B`. While being a correct solution in theory, since the user does state this when using this theorem. This is essentially not part of these functions, since they only set the velocity of the second point `P` and shouldn't be changing the velocity of `O`.
The solution I would suggest is to check whether the velocity of `O` in `B` is set to `0`. This check has 3 possible outcomes:
1. `O.vel(B)==0`, in which case you can just continue
2. `O.vel(B)!=0`, the user is making a mistake and an exception should occur
3. `O.vel(B)` gives a `ValueError` since the velocity is not specified yet. An option could be to make it more "user friendly" (personally I would just give this error back to the user) is to warn the user the velocity will be set to `0`.

However this is I think still not actually fully correct. Because as shown in the added image, the velocity theorems don't actually fully imply that point `P` should be fixed in `B`. In fact `P` and `Q` can both be moving points in the body, as long as they move with the same velocity, this is shown by putting the body `B` in the body `C`, where `S` is fixed. Note that frame `A` and `B` consist of the same unitvectors. So summarising you can say here that 


Hope this part is clear otherwise I can try to explain it a bit better

This also explains the "issues" shown. For example the correctness of `O.vel(B)` can be reasoned by putting the statements before in a logical order:
```py
# Define positions
Q.set_pos(O, l*A.y)
P.set_pos(Q, x*B.y)
# Following lines compute O im A and fully define it with respect to Q
O.set_vel(N, 0)
O.set_vel(A, 0)
Q.v2pt_theory(O, N, A)
# Following two define the velocity of Q in B
P.v1pt_theory(Q, N, B)
P.set_vel(B, x_dot*B.y)
# Since O is fully defined with respect to Q, which velocity is known in B,
# the velocity of O in B is also known
```

From this my conclusion would be that the current implementation is mathematically correct, however to be a bit more friendly to the user we could specify that the velocity of the reference point in the interframe should be known
proof
.

**Comment by moorepants** [comment]
I chatted with @TJStienstra (he's a student in my course here at TU Delft) and I favor his option #2. We should raise an error if you have previously set the velocity of P in B to some non-zero value when you call v2pt_theory with P as the first argument. P has to be fixed in B for the v2pt theorem to be valid.

**Comment by TJStienstra** [comment]
After implementing this I actually see that it gives quite a number of errors on other tests, where the velocity is not explicitly set to `0`. As well as some errors in other locations, namely:
- [models.py](https://github.com/sympy/sympy/blob/641a8b2839f0756808fb4cf941b82955f527a231/sympy/physics/mechanics/models.py#L200), where `points[-1].set_vel(Bi, 0)` should be added
- [joint.py](https://github.com/sympy/sympy/blob/641a8b2839f0756808fb4cf941b82955f527a231/sympy/physics/mechanics/joint.py#L531) which is related to #22956

Overall, implementing the check `O.vel(B)!=0` nicely shows, where possible problems are, but it will cause quite a lot of code (not only the code within the library) to break. This as can be seen above is sometimes pointing towards wrong assumptions, but most of the times it will just be annoying.

@moorepants
So should I create a PR implementing this constraint, of course solving the broken tests etc (excluding #22956, which is another issue)?
This problem of not checking whether `otherpoint` is actually fixed does not only apply for the velocity theorems, but also for the acceleration theorems (i.e. `a1pt_theory` and `a2pt_theory`). Should a constraint also be added there to check if `(otherpoint.vel(fixedframe)!=0 and otherpoint.acc(fixedframe)!=0)`?

**Comment by moorepants** [comment]
> So should I create a PR implementing this constraint, of course solving the broken tests etc (excluding https://github.com/sympy/sympy/issues/22956, which is another issue)?

Yes, it's best to open a PR now and we can all see what is going on.

> Should a constraint also be added there to check if (otherpoint.vel(fixedframe)!=0 and otherpoint.acc(fixedframe)!=0)?

Probably.

**Comment by moorepants** [comment]
One possible issue is that we think of these points that exist at a particular instance of time. Point T in this explanation: https://moorepants.github.io/learn-multibody-dynamics/translational.html#velocity-one-point-theorem

In equation 55 we apply the two point theorem to R but R is really T (a different point that is only at the same location as R at that instance of time). Maybe this will cause issues, because you can apply the two point theorem to R but only if you consider it as T (which is really just a thought experiment).
